source/src/microapp/cs_MicroappProtocol.cpp

```diff
@@ -30,574 +29,91 @@
 }
 
-void handlePinActionCommand(pin_cmd_t* pin_cmd) {
-	CommandMicroappPin pin = (CommandMicroappPin)pin_cmd->pin;
-	LOGi("Clear, set or configure pin %i", pin);
-	CommandMicroappPinOpcode2 opcode2 = (CommandMicroappPinOpcode2)pin_cmd->opcode2;
-	switch(opcode2) {
-		case CS_MICROAPP_COMMAND_PIN_WRITE: {
-			TYPIFY(EVT_GPIO_WRITE) gpio;
-			gpio.pin_index = pin;
-			CommandMicroappPinValue val = (CommandMicroappPinValue)pin_cmd->value;
-			switch(val) {
-				case CS_MICROAPP_COMMAND_VALUE_OFF: {
-					LOGi("Clear GPIO pin");
-					gpio.length = 1;
-					uint8_t buf[1];
-					buf[0] = 0;
-					gpio.buf = buf;
-					event_t event(CS_TYPE::EVT_GPIO_WRITE, &gpio, sizeof(gpio));
-					EventDispatcher::getInstance().dispatch(event);
-					break;
-				}
-				case CS_MICROAPP_COMMAND_VALUE_ON: {
-					LOGi("Set GPIO pin");
-					gpio.length = 1;
-					uint8_t buf[1];
-					buf[0] = 1;
-					gpio.buf = buf;
-					event_t event(CS_TYPE::EVT_GPIO_WRITE, &gpio, sizeof(gpio));
-					EventDispatcher::getInstance().dispatch(event);
-					break;
-				}
-				default:
-					LOGw("Unknown switch command");
-					return;
-			}
-			break;
-		}
-		case CS_MICROAPP_COMMAND_PIN_INPUT_PULLUP:
-		case CS_MICROAPP_COMMAND_PIN_READ: {
-			// TODO; (not that we do not handle event handler registration here but in SetMode above
-			break;
-		}
-		default:
-			LOGw("Unknown pin mode / opcode: %i", pin_cmd->opcode2);
-			return;
-	}
-}
+extern "C" {
 
-/*
- * Handle pin commands
- */
-void handlePinCommand(pin_cmd_t* pin_cmd) {
-	CommandMicroappPin pin = (CommandMicroappPin)pin_cmd->pin;
-	switch(pin) {
-		case CS_MICROAPP_COMMAND_PIN_SWITCH: { // same as DIMMER
-			handleSwitchCommand(pin_cmd);
-			break;
-		}
-		case CS_MICROAPP_COMMAND_PIN_GPIO1:
-		case CS_MICROAPP_COMMAND_PIN_GPIO2:
-		case CS_MICROAPP_COMMAND_PIN_GPIO3:
-		case CS_MICROAPP_COMMAND_PIN_GPIO4:
-		case CS_MICROAPP_COMMAND_PIN_BUTTON1:
-		case CS_MICROAPP_COMMAND_PIN_BUTTON2:
-		case CS_MICROAPP_COMMAND_PIN_BUTTON3:
-		case CS_MICROAPP_COMMAND_PIN_BUTTON4:
-		case CS_MICROAPP_COMMAND_PIN_LED1:
-		case CS_MICROAPP_COMMAND_PIN_LED2:
-		case CS_MICROAPP_COMMAND_PIN_LED3:
-		case CS_MICROAPP_COMMAND_PIN_LED4: {
-			CommandMicroappPinOpcode1 opcode1 = (CommandMicroappPinOpcode1)pin_cmd->opcode1;
-			// We have one virtual device at location 0, so we have to decrement the pin number by one to map
-			// to the array in the GPIO module
-			pin_cmd->pin--;
-			switch(opcode1) {
-				case CS_MICROAPP_COMMAND_PIN_MODE:
-					handlePinSetModeCommand(pin_cmd);
-					break;
-				case CS_MICROAPP_COMMAND_PIN_ACTION:
-					handlePinActionCommand(pin_cmd);
-					break;
-				default:
-					LOGw("Unknown opcode1");
-					return;
-			}
-			break;
-		}
-		default:
-			LOGw("Unknown pin: %i", pin_cmd->pin);
-	}
-}
+int microapp_callback(uint8_t* payload, uint16_t length) {
+	if (length == 0) return ERR_WRONG_PAYLOAD_LENGTH;
+	if (length > MAX_PAYLOAD) return ERR_WRONG_PAYLOAD_LENGTH;
```

Comment:
Always have brackets for if statements.
```suggestion
	if (length == 0 || length > MAX_PAYLOAD) {
		return ERR_WRONG_PAYLOAD_LENGTH;
	}
```